### DISCUSSION TITLE
Use lint setting or not ?

### DISCUSSION THREAD

**Discussion by wu-clan** [astral-sh/ruff#8297] [Q&A] [answered]
what is the difference between `[ruff]` and `[ruff.lint]` in the configuration? I don't seem to see documentation on this.

**Comment by tjkuson** [comment, marked as answer]
`[ruff]` is for configurations for the entire tool (such as files for Ruff to ignore), whereas `[ruff.lint]` is for configuring specifically the linter (such as which lint checks to enable). Similarly, `[ruff.format]` is for the Ruff formatter configurations, such as whether to use single or double quotes.

**Reply by charliermarsh** [reply]
Yeah, `[ruff.lint]` is relatively new and was introduced with the release of the formatter. We recommend it going forward, but the settings at the top-level `[ruff]` section are still respected for backwards compatibility.

For example, a setting like `line-length` that affects both the linter and formatter would go in `[ruff]`:

```toml
[ruff]
# Affects both tools...
line-length = 120

[ruff.lint]
# Only relevant for the linter...
select = ["E", "W", "F"]

[ruff.format]
# Only relevant for the formatter...
indent-style = "tab"
```

**Reply by MichaReiser** [reply]
To extend on @charliermarsh's response

We introduced `lint` when releasing the formatter beta to make it clarify the scope of a configuration and we're now looking for feedback before stabilizing (and promoting) `lint` over the top-level options.

**Reply by abravalheri** [reply]
Hi @charliermarsh, @MichaReiser, thank you very much for the clarification.
In a [recent discussion](https://github.com/pypa/setuptools/pull/4125/files#r1441597761) however, @DimitriPapadopoulos and I were wondering about if we should use the `[lint]` section or not.

Dimitri posted an interesting comment that make sense for me:

> They are going towards lint for all options that are related to lint, that's for sure. However, ignore applies to both lint and format. At least that's my understanding.

Is it the case that if we add a `lint` section, then the configurations will not apply for `format`?
Or will the `format` action just trigger if the `lint` tells the line needs fixing.

Going forward, should we use this logic (i.e., global configs apply to both `lint` and `format`) or should we always separate the configs in 2 different sections?

**Reply by abravalheri** [reply]
To be very specific:
If I have the following configuration:

```toml
# ruff.toml
[lint]
ignore = [
	# https://docs.astral.sh/ruff/formatter/#conflicting-lint-rules
	"W191",
	"E111",
	"E114",
	"E117",
	"D206",
	"D300",
	"Q000",
	"Q001",
	"Q002",
	"Q003",
	"COM812",
	"COM819",
	"ISC001",
	"ISC002",
]
```

Is the most up-to-date recommendation for it to be inside the `[lint]` section? And is it the case the "global" `ignore` is just there fore backward compatibility?

Or is it the case that the "global" `ignore` behaves differently than `lint.ignore` and therefore I should use the "global" ignore because it dictates how both `lint` and `format` works?

**Reply by charliermarsh** [reply]
@abravalheri - Yup, what you have there is correct and recommended. The "global" `ignore` exists for backwards compatibility and will be removed at some point in the future.

**Reply by charliermarsh** [reply]
There are some settings that exist under `lint` for which this is _not_ true. For example, `line-length` can be set at the top-level, or under `lint` and/or under `format`, for cases in which you need separate line length enforcement for linting and formatting. And there are some settings that always affect both, like `target-version`. But in the majority of cases, if a setting is supported under `lint`, that's where you should put it.

**Reply by abravalheri** [reply]
Thank you very much @charliermarsh!

**Reply by DimitriPapadopoulos** [reply]
@charliermarsh The current documentation shows us all [**Top-level** ](https://docs.astral.sh/ruff/settings/#top-level) settings such as [`ignore`](https://docs.astral.sh/ruff/settings/#ignore) at the tool root level, i.e.:

* **pyproject.toml**
  ```toml
  [tool.ruff]
  # Skip unused variable rules (`F841`).
  ignore = ["F841"]
  ```
* **ruff.toml**
  ```toml
  # Skip unused variable rules (`F841`).
  ignore = ["F841"]
  ```

Should the documentation change? Current documentation should show the currently recommended location under `[lint]`, not the backwards compatible "global" location that may become obsolete in the future.

**Reply by MichaReiser** [reply]
I'm sorry for the confusion that the `lint` / `format` sections have created. The idea was to release them as experimental and collect feedback before making them the recommended configuration option. That's why the documentation defaults to the global options today (because `lint` is still considered experimental).

I think we should make the switch over soon, but keeping backward compatibility by changing the documentation to recommend the usage of `lint` and remove the `experimental` note from the `lint` section.

**Reply by DimitriPapadopoulos** [reply]
No worries, it all works well for us.

I have a related question though. In the long term, will you remove the possibility to have a top-level [exclude](https://docs.astral.sh/ruff/settings/#exclude) setting?
> [`exclude`](https://docs.astral.sh/ruff/settings/#exclude) [#](https://docs.astral.sh/ruff/settings/#exclude)
> 
> A list of file patterns to exclude from formatting and linting.

It might make sense to keep some settings at top-level. Or perhaps not to avoid confusion :smile: I am asking because I had to duplicate the `exclude` setting, which is shared by the linter and the formatter in this specific project.

**Reply by MichaReiser** [reply]
We'll keep the top-level `exclude` setting. The idea is that:

* The top-level `exclude`: Configures the files of your entire project
* Lint/Formatter level: Configure the files to exclude for each individual tool
